lib/rucio/core/monitor.py

```diff
@@ -35,27 +44,88 @@
+    In case of statsd, metrics are formatted using str.format(**labels). The prometheus
+    ones using metric.labels(**labels) calls.
+
+    If the prometheus metric string is not provided, it is derived from the statsd one.
+    """
+
+    def __init__(self, statsd, prom=None, documentation=None, labelnames=(), labelvalues=None, registry=None):
+        """
+        :param statsd: a string, eventually with keyword placeholders for the str.format(**labels) call
+        :param prom: a string or a prometheus metric object
+        """
+        self._registry = registry or REGISTRY
+        self._documentation = documentation or ''
+        self._statsd = statsd
+        if not prom:
+            # automatically generate a prometheus metric name
+            #
+            # remove '.{label}' from the string for each `label`
+            stats_without_labels = ''.join(tup[0].rstrip('.') for tup in string.Formatter().parse(self._statsd))
+            prom = 'rucio_{}'.format(stats_without_labels).replace('.', '_')
+        if isinstance(prom, str):
+            self._prom = self.init_prometheus_metric(prom, self._documentation, labelnames=labelnames, labelvalues=labelvalues)
+        else:
+            self._prom = prom
+        self._labelnames = labelnames
+        self._labelvalues = labelvalues
+
+    @abstractmethod
+    def init_prometheus_metric(self, name, documentation, labelnames=(), labelvalues=None):
+        pass
+
+    def labels(self, **labelvalues):
+        return self.__class__(
+            prom=self._prom.labels(**labelvalues),
+            statsd=self._statsd.format(**labelvalues),
+            documentation=self._documentation,
+            labelnames=self._labelnames,
+            labelvalues=labelvalues,
+            registry=self._registry,
+        )
+
+
+class MultiCounter(MultiMetric):
+
+    def inc(self, delta=1):
+        self._prom.inc(delta)
+        CLIENT.incr(self._statsd, delta)
 
+    def init_prometheus_metric(self, name, documentation, labelnames=(), labelvalues=None):
+        return Counter(name, documentation, labelnames=labelnames, labelvalues=labelvalues, registry=self._registry)
 
-def record_counter(counters, delta=1):
+
+def record_counter(name, delta=1, labels=None):
     """
     Log one or more counters by arbitrary amounts
 
-    :param counters: The counter or a list of counters to be updated.
+    :param name: The counter to be updated.
     :param delta: The increment for the counter, by default increment by 1.
+    :param labels: labels used to parametrize the metric
     """
-    if isinstance(counters, list):
-        for counter in counters:
-            if delta > 0:
-                CLIENT.incr(counter, delta)
-            else:
-                CLIENT.decr(counter, delta)
+
+    counter = COUNTERS.get(name)
+    if not counter:
+        COUNTERS[name] = counter = MultiCounter(statsd=name, labelnames=labels.keys() if labels else ())
+
+    if delta < 0:
```

Comment:
do you mean here:

`delta = abs(delta)` ?

Reply:
Absolutely correct. The code which is being substituted contained an "if delta > 0". Because of this, my dumb brain didn't realize it can be written differently :D

Reply:
:-D